parachain/pallets/basic-channel/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+//! # Basic Channel
+//!
+//! The Basic Channel module is is a non-incentivized bridge between Ethereum and Pokadot
+//! ecosystems. It is meant to be a low-barrier entry to cross-chain communication without
+//! guarantees, but also without fees.
+//!
+//! ## Implementation
+//!
+//! Before a [`Message`] is dispatched to a target [`Application`], it is submitted to a [`Verifier`] for verification.
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Calls
+//!
+//! - `submit`: Submit a message for verification and dispatch.
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult};
+use frame_system::{self as system, ensure_signed};
+
+use artemis_core::{
+	Message, MessageCommitment, MessageDispatch, MessageId, SubmitOutbound, Verifier,
+};
+use channel::{inbound::BasicInboundChannel, outbound::BasicOutboundChannel};
+use envelope::Envelope;
+use primitives::{InboundChannel, InboundChannelData, OutboundChannel, OutboundChannelData};
+use sp_core::H160 as EthAddress;
+use sp_std::convert::TryFrom;
+use sp_std::prelude::*;
+
+mod channel;
+mod envelope;
+#[cfg(test)]
+mod mock;
+mod primitives;
+#[cfg(test)]
+mod tests;
+
+type MessageNonce = u64;
+
+pub trait Config: system::Config {
+	type Event: From<Event<Self>> + Into<<Self as system::Config>::Event>;
+
+	/// Verifier module for message verification.
+	type Verifier: Verifier;
+
+	/// Used by outbound channels to persist messages for outbound delivery.
+	type MessageCommitment: MessageCommitment;
+
+	/// Verifier module for message verification.
+	type MessageDispatch: MessageDispatch<MessageId>;
+}
+
+decl_storage! {
```

Comment:
Now that we've split out the basic channel from the bridge, some of the abstractions and associated boilerplate in the old design no longer make sense.

Particularly, we can just store the account nonces directly, without putting them into those `InboundChannelData` or `OutboundChannelData` structs.

For example
```rust
decl_storage! {
	trait Store for Module<T: Config> as BasicChannelModule {
		/// Nonces for inbound channels.
		pub InboundNonces: map hasher(identity) H160 => u64;
		/// Storage for outbound channels.
		pub OutboundNonces: map hasher(identity) T::AccountId => u64;
	}
}
```

By implication we can mostly just remove all the code under `channels/`, except for the `InboundChannel::submit` and `OutboundChannel::submit` methods. We should move those methods to a impl block like this:

```
impl<T: Config> Module<T> {
  fn submit_inbound(...) { ... }
  fn submit_outbound(...) { ... }
}
```

Reply:
Yeah totally, I was thinking of it too.
The only thing that would prevent it is if we need something more complex for a (user, channel) -> nonce mapping, but I was thinking on using double maps for that (need to look further)